src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -130,7 +140,20 @@ public Task InitializeAsync(ProtectedPrerenderComponentApplicationStore store, C
                 // At this point all components have successfully produced an initial render and we can clear the contents of the component
                 // application state store. This ensures the memory that was not used during the initial render of these components gets
                 // reclaimed since no-one else is holding on to it any longer.
-                store.ExistingState.Clear();
+                // This is also important because otherwise components will keep reusing the existing state after
+                // the initial render instead of initializing their state from the original sources like the Db or a
+                // web service, preventing UI updates.
+                if (Descriptors.Count > 0)
+                {
+                    store.ExistingState.Clear();
+                }
+
+                // This variable is used to track that this is the first time we are updating components.
+                // In Blazor Web scenarios the app will send an initial empty list of descriptors,
+                // so we want to make sure that we allow setting up the state in that case.
+                // In Blazor Server the initial set of descriptors is provided via the call to Start, so
+                // we want to make sure we don't take any state afterwards.
+                _isFirstUpdate = Descriptors.Count == 0;
```

Comment:
What if it goes back down to zero later, after the first render?

Reply:
It can't. The descriptors are populated at the time the circuit is created and no longer updated (it's a readonly list). That's essentially what I am using to distinguish between old blazor server apps and blazor web.

Reply:
Once you provide an initial set of descriptors via a call to "StartCircuit" you are locked on to "Blazor Server" behaviors and vice versa.